### DISCUSSION TITLE
Finding entries matching at least one tag

### DISCUSSION THREAD

**Discussion by LucasBW** [prisma/prisma#23188] [Q&A]
### Question

Hello, I have a prisma schema like this: 

```prisma
model Post {
  id    Int        @id @default(autoincrement())
  title String
  tags  PostTags[]
}

model PostTags {
  id     Int   @id @default(autoincrement())
  post   Post? @relation(fields: [postId], references: [id])
  tag    Tag?  @relation(fields: [tagId], references: [id])
  postId Int?
  tagId  Int?
  @@unique['postId ','tagId']
}

model Tag {
  id    Int        @id @default(autoincrement())
  name  String     @unique
  posts PostTags[]
}
```

The table PostTags contains:
---
| postId| tagId  | 
| ---: | ---: | 
| 18 | 1 | 
| 19 | 2 | 
| 19 | 3 | 
| 10 | 3 | 
| 18 | 3 | 
| 18 | 4 | 
| 18 | 7 | 
| 19 | 7 | 

I'm trying to find all postId that are matching at least a tagId in an array. 

**As example:** 
**One ID:**
`tagIds = [3]`

Result would be: 
PostTags 
---
| postId|
| ---: | 
| 19 |
| 10 |
| 18 |

**Two ID:**
`tagIds = [3,7]`

Result would be: 
PostTags 
---
| postId| 
| ---: | 
| 18 |
| 19 | 

**Three ID:**
`tagIds = [3,7,1]`

Result would be: 
PostTags 
---
| postId| 
| ---: | 
| 18 |

The logic would be to find all posts matching minimum all tag setup in the filter but can have more (but no less).

In raw SQL I found something like this: 
```sql
SELECT "postId"
FROM "PostTags "
WHERE "tagId"  IN (3, 7)
GROUP BY "postId"
HAVING COUNT(DISTINCT "tagId") >= 2;
```

But I was wondering, is there a way to do that through prisma without a raw query ? 

### Information about Prisma Schema, Client Queries and Environment (optional)

- Database: PostgreSQL
prisma                       : 5.8.0
@prisma/client          : 5.8.0

**Comment by ludralph** [comment]
Hi @LucasBW 👋 

Thank you for raising this question.

You can use the `findMany` method with the `where` and `in` operators to filter records based on an array of values.
Here is an example of how you can use `findMany` to filter `PostTags` based on an array of `tagId`:
```js
const tagIds = [3, 7]; // your array of tagIds

const postTags = await prisma.postTags.findMany({
  where: {
    tagId: {
      in: tagIds,
    },
  },
});
```
This will return all `PostTags` where `tagId` is in the `tagIds` array. However, this does not guarantee that the returned `PostTags` will have all the `tagIds` in the array, which is what the SQL `HAVING` clause does.

To achieve the same result as the SQL `HAVING` clause, you would need to process the results in your application code. You could do this by grouping the `PostTags` by `postId` and then filtering the groups based on the count of distinct tagId.
```js
// Group PostTags by postId
const groupedPostTags = postTags.reduce((groups, postTag) => {
  const key = postTag.postId;
  if (!groups[key]) {
    groups[key] = [];
  }
  groups[key].push(postTag);
  return groups;
}, {});

// Filter groups based on the count of distinct tagId
const filteredGroups = Object.entries(groupedPostTags).filter(([postId, postTags]) => {
  const distinctTagIds = [...new Set(postTags.map(postTag => postTag.tagId))];
  return distinctTagIds.length >= tagIds.length;
});

// Get the postId of the filtered groups
const filteredPostIds = filteredGroups.map(([postId]) => postId);
```

In this example, `filteredPostIds` will contain the `postId` of all `PostTags` that have all the `tagIds` in the array.

Please note that this is not as efficient as doing it directly in the database with a `SQL HAVING` clause, especially if you have a large number of `PostTags`. If performance is a concern, you might want to consider using  `prisma.$queryRaw` method to execute it.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
